File: src/SWTrainController/Controller.cpp
//
// Created by Collin Hough on 10.7.20
// Controller implementation file
//

#include "Controller.h"

/**
 * @brief initialize variables
 */
Controller::Controller()
{
    // Initialize vital variables
    command_speed = 0;
    current_speed = 0;
    setpoint_speed = 0;
    power_command = 0;
    authority = 0;
    mode = 0;
    serviceBrake = 0;
    emergencyBrake = 0;
    kp = 0;
    ki = 0;
    uk = 0;
    uk1 = 0;
    ek = 0;
    ek1 = 0;
    signalPickupFailure = 0;
    engineFailure = 0;
    brakeFailure = 0; 
    NonVitalOperations();
}

/** 
 * @brief initializes data coming from train model
 * @param com_sp = command speed
 * @param curr_sp = current speed
 * @param auth = authority
 */
Controller::Controller(float com_sp, float curr_sp, bool auth)
{
    command_speed = com_sp;
    current_speed = curr_sp;
    setpoint_speed = 0;
    power_command = 0;
    authority = auth;
    mode = 0;
    serviceBrake = 0;
    emergencyBrake = 0;
    kp = 0;
    ki = 0;
    uk = 0;
    uk1 = 0;
    ek = 0;
    ek1 = 0;
    signalPickupFailure = 0;
    engineFailure = 0;
    brakeFailure = 0;
    NonVitalOperations();
}

///////////////////////////////////////////////////////////////
// SETTERS AND GETTERS
///////////////////////////////////////////////////////////////

/**
 * @brief Setter function for Kp
 * @param KP = kp
 */
void Controller::setKp(float KP)
{
    kp = KP;
}

/**
 * @brief Setter function for Ki
 * @param KI = ki
 */
void Controller::setKi(float KI)
{
    ki = KI;
}

/**
 * @brief Setter function for command speed
 * @param com_sp == command_speed
 */
void Controller::setCommandSpeed(float com_sp)
{
    command_speed = com_sp;
}

/**
 * @brief Setter function for current speed
 * @param curr_sp == current_speed
 */
void Controller::setCurrentSpeed(float curr_sp)
{
    current_speed = curr_sp;
}

/**
 * @brief Setter function for setpoint speed
 * @param setp_sp == setpoint speed
 */
void Controller::setSetpointSpeed(float setp_sp)
{
    setpoint_speed = setp_sp;
}

/**
 * @brief Setter function for authority
 * @param auth == auth
 */
void Controller::setAuthority(bool auth)
{
    authority = auth;
}

/**
 * @brief Getter function for command speed
 * @return command_speed
 */
float Controller::getCommandSpeed()
{
    return command_speed;
}

/**
 * @brief Getter function for current speed
 * @return current_speed
 */
float Controller::getCurrentSpeed()
{
    return current_speed;
}

/**
 * @brief Getter function for setpoint speed
 * @return setpoint speed
 */
float Controller::getSetpointSpeed()
{
    return setpoint_speed;
}

/**
 * @brief Getter function for authority
 * @return authority
 */
bool Controller::getAuthority()
{
    return authority;
}

/**
 * @brief Getter function for power command
 */
float Controller::getPowerCommand()
{
    return power_command;
}

/**
 * @brief Getter function for mode
 */
bool Controller::getMode()
{
    return mode;
}

/**
 * @brief Getter function for service brake
 */
bool Controller::getServiceBrake()
{
    return serviceBrake;
}

// Non-Vital Getters
/**
 * @brief open/close doors
 */
bool Controller::getDoors()
{
    return NVO.doors;
}

/**
 * @brief turn lights on/off
 */
bool Controller::getLights()
{
    return NVO.lights;
}

/**
 * @brief turn announcements on/off
 */
bool Controller::getAnnounceStations()
{
    return NVO.announcements;
}

/**
 * @brief turn advertisements on/off
 */
bool Controller::getAds()
{
    return NVO.advertisements;
}

///////////////////////////////////////////////////////////////
// VITAL OPERATIONS
///////////////////////////////////////////////////////////////
/**
 * @brief calculates power command that will be sent to train model
 */
void Controller::calculatePower()
{
    // Find Verror depending on mode
    float Verror = 0;
    if (mode == 0) // Automatic Mode
        Verror = command_speed - current_speed;
    else // Manual Mode
        Verror = setpoint_speed - current_speed;

    // Set ek as the kth sample of velocity error
    ek = Verror;

    // Determine uk as shown in slide 65 of lecture 2
    if (power_command < MAX_POWER)
    {
        // Find uk 
        uk = uk1 + (T/2) * (ek + ek1);
    }
    else
    {
        uk = uk1;
    }
    
    // Find power command
    power_command = (kp * ek) + (ki * uk);

    // Set past values of uk and ek
    uk1 = uk;
    ek1 = ek;

}

/**
 * @brief ensures train does not exceed speed limit
 * @brief if train is in automatic mode, speed is set to command speed
 * @brief if train is in manual mode, speed is set to setpoint speed
 */
void Controller::regulateSpeed()
{
    // Check mode of operation (Automatic or Manual)
        // Automatic Mode
            // C
        // Manual Mode
}

/**
 * @brief safety critical aspect to stop train immediately
 */
void Controller::activateEmergencyBrake()
{
    emergencyBrake = 1;
}

/**
 * @brief safety critical aspect to reset emergency brake
 */
void Controller::resetEmergencyBrake()
{
    emergencyBrake = 0;
}

/**
 * @brief toggle service brake on and off
 */
bool Controller::toggleServiceBrake()
{
    serviceBrake = !serviceBrake;
    return serviceBrake;
}

/**
 * @brief allows operator to switch between manual and automatic mode
 * @param override = string code entered by operator to initiate manual override
 * @return returns boolean value to signify success of operation
 */
bool Controller::toggleMode(std::string override)
{
    // Check if override code is correct
    if (override == password)
    {
        mode = !mode;
        return mode;
    }
    else
        return 0;
}

///////////////////////////////////////////////////////////////
// NON-VITAL OPERATIONS
///////////////////////////////////////////////////////////////
/**
 * @brief open/close doors
 */
bool Controller::toggleDoors()
{
    NVO.doors = !NVO.doors;
    return NVO.doors;
}

/**
 * @brief turn lights on/off
 */
bool Controller::toggleLights()
{
    NVO.lights = !NVO.lights;
    return NVO.lights;
}

/**
 * @brief turn announcements on/off
 */
bool Controller::announceStations()
{
    NVO.announcements = !NVO.announcements;
    return NVO.announcements;
}

/**
 * @brief turn advertisements on/off
 */
bool Controller::toggleAds()
{
    NVO.advertisements = !NVO.advertisements;
    return NVO.advertisements;
}

/**
 * @brief sets temperature of train cabin
 * @param temp = temperature cabin is set to
 */
void Controller::setCabinTemp(float temp)
{
    NVO.temperature = temp;
}

/**
 * @brief gets temperature of train cabin
 * @return returns temperature of cabin
 */
float Controller::getCabinTemp()
{
    return NVO.temperature;
}

File: src/SWTrainController/Controller.py
"""
    Author: Collin Hough

    Date: 10.7.20
"""
from src.common_def import Converters

# Defines the maximum power of the train engine
MAX_POWER = 120000 # Units for max power are kW

# Define automatic to manual override password
PASSWORD = "override"

# Define sampling period for calculating power (1/10th of a second)
T = 0.2

class Controller:
    """ Defines controller to be used on each train """

    def __init__(self, com_sp = 0, curr_sp = 0, auth = 0):
        """
            Initializes data coming from train model
            com_sp = command speed
            curr_sp = current speed
            auth = authority
        """
        # Safety critical information
        self.command_speed = com_sp
        self.current_speed = curr_sp
        self.setpoint_speed = 0.0
        self.power_command = 0.0
        self.authority = auth
        self.mode = False # 0 = Automatic, 1 = Manual
        self.service_brake = True
        self.emergency_brake = False

        # Train Engineer inputs
        self.kp = 0.0
        self.ki = 0.0

        # Variables for power calculation
        self.uk = 0.0
        self.uk1 = 0.0
        self.ek = 0.0
        self.ek1 = 0.0

        # NonVital Operations (0 = ON, 1 = OFF)
        self.doors = 0
        self.announcements = 0
        self.lights = 0
        self.temperature = 70
        self.advertisements = 0

        # Failure cases
        self.signal_pickup_failure = False
        self.engine_failure = False
        self.brake_failure = False

        # Used for starting and stopping the train's power loop
        self.hold_power_loop = False

    ###############################/
    # VITAL OPERATIONS
    ###############################/

    def calculate_power(self):
        """
            Calculates power command that will be sent to train model
            Units:
                Kp: (W / (m/s) )
                Ki: (W / m)
                Command Speed: m/s
                Current Speed: m/s
                Setpoint Speed: m/s
                ek/ek1: m/s
                uk/uk1: m
        """
        # Convert Command Speed, Current Speed, and Setpoint Speed into m/s
        self.convert_command_speed = self.command_speed * Converters.KmHr_to_mps
        self.convert_current_speed = self.current_speed * Converters.MPH_to_mps
        self.convert_setpoint_speed = self.setpoint_speed * Converters.MPH_to_mps

        # Find Verror depending on mode
        velocity_error = 0
        if self.mode == 0: # Automatic Mode
            velocity_error = self.convert_command_speed - self.convert_current_speed
        else: # Manual Mode
            velocity_error = self.convert_setpoint_speed - self.convert_current_speed

        # Set ek as the kth sample of velocity error
        self.ek = velocity_error

        # Determine uk as shown in slide 65 of lecture 2
        if self.power_command < MAX_POWER:
            # Find uk
            self.uk = self.uk1 + (T/2) * (self.ek + self.ek1)
        else:
            self.uk = self.uk1

        # Check for engine failure
        if self.power_command > 0 and self.current_speed == 0:
            self.engine_failure == True

        # Find power command
        if self.service_brake == True or self.emergency_brake == True:
            self.power_command = 0
            self.uk = 0
            self.ek = 0
        else:
            self.power_command = (self.kp * self.ek) + (self.ki * self.uk)
            # Cut off power at appropriate limits
            if(self.power_command > 120000):
                self.power_command = 120000
            elif(self.power_command < -120000):
                self.power_command = -120000

        # Set past values of uk and ek
        self.uk1 = self.uk
        self.ek1 = self.ek


    # @brief ensures train does not exceed speed limit
    # @brief if train is in automatic mode, speed is set to command speed
    # @brief if train is in manual mode, speed is set to setpoint speed
    # void regulateSpeed() IDK IF I NEED THIS

    def activate_emergency_brake(self):
        """ Safety critical aspect to stop train immediately """
        self.emergency_brake = True

    def reset_emergency_brake(self):
        """ Safety critical aspect to reset emergency brake """
        self.emergency_brake = False

    def toggle_service_brake(self):
        """ Toggle service brake on and off """
        # Check for potential failure
        if self.service_brake == True:
            if self.brake_failure == True:
                # If brake failure occurs do not change service brake
                self.service_brake = True
            else:
                # If no brake failure toggle brake normally
                self.service_brake = not self.service_brake
        else:  
            self.service_brake = not self.service_brake

    def toggle_mode(self, override):
        """ Allows operator to switch between manual and automatic mode """
        # Check if override code is correct
        if override == PASSWORD:
            self.mode = not self.mode

    ###############################/
    # NON-VITAL OPERATIONS
    ###############################/

    def toggle_doors(self):
        """ Open/close doors """
        self.doors = not self.doors

    def toggle_lights(self):
        """ Turn lights on/off """
        self.lights = not self.lights

    def toggle_announcements(self):
        """ Turn announcements on/off """
        self.announcements = not self.announcements

    def toggle_ads(self):
        """ Turn advertisements on/off """
        self.advertisements = not self.advertisements
